third_party/move/move-vm/runtime/src/storage/loader/native.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    storage::loader::traits::StructDefinitionLoader, ModuleStorage, RuntimeEnvironment,
+    WithRuntimeEnvironment,
+};
+use move_binary_format::errors::PartialVMResult;
+use move_vm_types::{
+    gas::{GasMeter, ModuleTraversalContext},
+    loaded_data::{runtime_types::StructType, struct_name_indexing::StructNameIndex},
+};
+use std::sync::Arc;
+
+/// Loader implementation used for natives and function value serialization. Prior to lazy loading,
+/// natives were not metering gas for modules. With lazy loading:
+///   1. For metering layout construction in native context. Because natives do not have access to
+///      gas meter trait, lazy loader currently pre-loads the transitive closure for that native
+///      and its type arguments. Hence, all layouts in native context must be fully metered.
+///   2. For function value serialization. Function values may capture arguments, and if stored to
+///      the global storage, layouts are constructed for these captured arguments. Note that we
+///      expect module used for layouts to be already loaded (because construction of captured
+///      arguments implies their layouts are loaded as well).
+// TODO(lazy-loading):
+//   It seems that (2) does not hold with enums though. If captured argument comes from the stack,
+//   some A::V2 for example, modules for A::V1 will not be loaded, but they will be accessed during
+//   layout construction... It seems that what "saves" us here is that we will check the depth of
+//   a value when we pack an enum and thus, load the modules used for layouts. Double check this.
+pub struct NativeLoader<'a> {
+    module_storage: &'a dyn ModuleStorage,
+}
+
+impl<'a> NativeLoader<'a> {
+    /// Returns a new native loader.
+    pub fn new(module_storage: &'a dyn ModuleStorage) -> Self {
+        Self { module_storage }
+    }
+}
+
+impl<'a> WithRuntimeEnvironment for NativeLoader<'a> {
+    fn runtime_environment(&self) -> &RuntimeEnvironment {
+        self.module_storage.runtime_environment()
+    }
+}
+
+impl<'a> StructDefinitionLoader for NativeLoader<'a> {
+    fn is_lazy_loading_enabled(&self) -> bool {
+        self.runtime_environment().vm_config().enable_lazy_loading
+    }
+
+    fn load_struct_definition(
+        &self,
+        _gas_meter: &mut impl GasMeter,
+        traversal_context: &mut dyn ModuleTraversalContext,
+        idx: &StructNameIndex,
+    ) -> PartialVMResult<Arc<StructType>> {
+        let struct_name = self
+            .runtime_environment()
+            .struct_name_index_map()
+            .idx_to_struct_name_ref(*idx)?;
+
+        if self.is_lazy_loading_enabled() {
```

Comment:
Can we add a trivial unit test (if it's trivial to add due to types), that checks the correct check errors is lazy loading is enabled (vs no error when special and loaded).

BTW why can't we do the check regardless of whether lazy loading is enabled?